### PR TITLE
Dynamic URL feature

### LINKED ISSUE #63: Dynamic short URL
## Description

Add dynamic URL type, so we can go to specified URL path dynamically.
Example:
- URL: somelongdomainname.com/{param}/view
- Shorten: ksana.in/slug/blahblah
- Destination: somelongdomainname.com/blahblah/view

Not sure if this are make sense and possible, but pretty sure this will be useful.

## Implementation Model

- Add `type` field  (enum: normal, dynamic) or add `dynamic` field (bool)
- If type is _dynamic_ then the access URL should have 2 parameters: `ksana.in/slug/paramvalue` otherwise it will 404 or something
- If type is _dynamic_ then replace `{param}` with given param in this case `blahblah` before redirecting

If we agree to add this feature i would happy to contribute.

### PR DESCRIPTION
**Pull request by bramaudi**
Closes #63 

## Description

This PR are for dynamic URL feature proposal.
Dynamic URL are feature to give custom parameter in our short URL to access specified website URL with dynamic URL query / path.

Currently multiple param or nested path (param contains slash "`/`") are not supported.

How its look compared to regular link:
| | Regular | Dynamic |
|-|------------|-------------|
| **Short URL** | ksana.in/slug | ksana.in/slug/param |
| **Destination** | domain.com | domain.com/param |

## What inside this PR

- [x] Add new field to database
- [x] Add option to form with validator
- [x] Update README

### CODE REVIEW

**Comment by bramaudi** [comment]
How to use:

https://user-images.githubusercontent.com/14371119/158523043-ad87f954-c0da-4dfe-a357-f74ed7e369fe.mp4

Demo: 

https://user-images.githubusercontent.com/14371119/158523078-001237f8-16d3-47ff-a42b-e7851e35729f.mp4

File: README.md [hunk 1]
```diff
@@ -57,6 +57,7 @@ create table urls (
   real_url text check (char_length(real_url) > 1),
   slug text check (char_length(slug) > 1),
   hit integer default 0,
+  is_dynamic boolean default false,
```
Comment (mazipan) [inline README.md:60]: Tipe datanya `smallint` aja, nanti diisi `1/0` , default `0`
`1` = `true`
`0` = `false`

```suggestion
  is_dynamic smallint default 0,
```

File: src/pages/[slug]/[param].tsx [hunk 2]
```diff
@@ -33,9 +39,12 @@ export async function getServerSideProps({ params }: IGetServerSideProps) {
       .update({ hit: data.hit + 1 })
       .match({ slug: slug })
 
+    const destination = data.is_dynamic ? data.real_url.replace(/{param}/, param) : data.real_url
+
+    console.log('params', params)
```
Comment (mazipan) [inline src/pages/[slug]/[param].tsx:44]: Remove log

File: README.md (hunk 1 above)
Comment (bramaudi) [reply README.md:60]: Benar juga mas, nilai lebih efisien :+1:

File: README.md [hunk 3]
```diff
@@ -57,6 +57,7 @@ create table urls (
   real_url text check (char_length(real_url) > 1),
   slug text check (char_length(slug) > 1),
   hit integer default 0,
+  is_dynamic smallint default 0,
```
Comment (mazipan) [inline README.md:60]: seems `supabase` have `int2` instead, can we use `int2`?

**Review by mazipan — APPROVED** [review]
Overall LGTM

File: src/pages/api/urls/save/[user_id].ts [hunk 4]
```diff
@@ -34,6 +34,7 @@ export default async (req: NextApiRequest, res: NextApiResponse) => {
         {
           real_url: url,
           slug: sanitizeSlug(slug),
+          is_dynamic: ~~is_dynamic,
```
Comment (mazipan) [inline src/pages/api/urls/save/[user_id].ts:37]: How if we simplify the code? `~~` seems not familiar enough for most of developers.

```suggestion
          is_dynamic: is_dynamic ? 1 : 0,
```

Comment (bramaudi) [reply src/pages/api/urls/save/[user_id].ts:37]: That's are simplified already IMO, but yeah it's not obvious compared to ternary condition. I'll change it.

File: README.md (hunk 3 above)
Comment (bramaudi) [reply README.md:60]: Not sure but `tinyint` are already showed as `int2` in Supabase Studio:

image

**Review by mazipan — APPROVED** [review]
Let's ship 🚢

**Pull request merged**